### PR TITLE
[EHealth] Add examples of NER and SPO tasks

### PR DESCRIPTION
**Pull request by LemonNoel**
### PR types

New features

### PR changes

Models

### Description

- Fix bug in NER example
- Add SPO example
- Add results in README

### CODE REVIEW

File: examples/biomedical/cblue/README.md [hunk 1]
```diff
@@ -56,7 +53,12 @@ $ python -m paddle.distributed.launch --gpus "0,1,2,3" train.py --dataset CHIP-C
 | KUAKE-QQR |   16   |     32     |      6e-5     |       64       | 0.82364 |
 | KUAKE-QTR |   12   |     32     |      6e-5     |       64       | 0.69653 |
 | KUAKE-QIC |    4   |     32     |      6e-5     |      128       | 0.81176 |
+| CMeEE     |    2   |     32     |      6e-5     |      128       | 0.66167 |
 
+```shell
+$ unset CUDA_VISIBLE_DEVICES
+$ python -m paddle.distributed.launch --gpus "0,1,2,3" train_classification.py --dataset CHIP-CDN-2C --batch_size 256 --max_seq_length 32 --learning_rate 3e-5 --epochs 16
```
Comment (wawltor) [inline examples/biomedical/cblue/README.md:61]: 这块的训练需要是多卡吗？ 单卡是否可以了？很多用户可能没有多卡，容易形成误导

File: examples/biomedical/cblue/train_ner.py [hunk 2]
```diff
@@ -148,7 +143,7 @@ def do_train():
 
     criterion = paddle.nn.functional.softmax_with_cross_entropy
 
-    metrics = [ChunkEvaluator(label_list[0]), ChunkEvaluator(label_list[1])]
+    metric = NEREvaluator(label_list)
```
Comment (wawltor) [inline examples/biomedical/cblue/train_ner.py]: 这块NER 是不是改成CBlueChunkEvaluator之类的

File: examples/biomedical/cblue/train_ner.py [hunk 3]
```diff
@@ -197,13 +187,14 @@ def do_train():
...
-                           f1, args.logging_steps / time_diff))
+                        'global step %d, epoch: %d, batch: %d, loss: %.5f, loss symptom: %.5f, loss others: %.5f, f1: %.5f, speed: %.2f step/s, learning_rate: %f'
+                        % (global_step, epoch, step, loss, losses[1], losses[0],
+                           f1, args.logging_steps / time_diff,
+                           lr_scheduler.get_lr()))
                     tic_train = time.time()
 
                 if global_step % args.valid_steps == 0 and rank == 0:
-                    evaluate(model, criterion, metrics, dev_data_loader)
+                    evaluate(model, criterion, metric, dev_data_loader)
```
Comment (wawltor) [inline examples/biomedical/cblue/train_ner.py:192]: 下面的代码，os.pthj -> os.path

File: examples/biomedical/cblue/utils.py [hunk 4]
```diff
@@ -164,6 +247,83 @@ def create_batch_label(ent_labels, spo_labels, num_classes, max_batch_len):
     return pad_ent_labels, pad_spo_labels
 
 
+class NEREvaluator(paddle.metric.Metric):
```
Comment (wawltor) [inline examples/biomedical/cblue/utils.py]: 同上，这块加一下document，同时说明一下和ChunckEvaluator的区别点

File: examples/biomedical/cblue/utils.py [hunk 5]
```diff
@@ -41,6 +44,52 @@ def create_dataloader(dataset,
...
+                None, it means initial learning rate. 
+                Defaults to -1.
+            verbose (bool, optional):
+                If True, prints a message to stdout for each update.
+                Defaults to False.
+        """
+
+        def is_integer(number):
+            if sys.version > '3':
+                return isinstance(number, int)
```
Comment (wawltor) [inline examples/biomedical/cblue/utils.py]: 这里不用考虑，目前paddleNLP的版本就是要求python 3.x

File: examples/biomedical/cblue/train_ner.py [hunk 6]
```diff
@@ -44,35 +43,31 @@ def set_seed(seed):
...
     losses = []
     for batch in data_loader:
         input_ids, token_type_ids, position_ids, masks, label_oth, label_sym = batch
-        logits = model(input_ids, token_type_ids, position_ids)
-        loss_oth = criterion(logits[0], paddle.unsqueeze(label_oth, 2))
-        loss_oth = paddle.mean(loss_oth * paddle.unsqueeze(masks, 2))
-        loss_sym = criterion(logits[1], paddle.unsqueeze(label_sym, 2))
-        loss_sym = paddle.mean(loss_sym * paddle.unsqueeze(masks, 2))
+        att_mask = paddle.scale(
+            masks, scale=10000.0, bias=-1.0, bias_after_scale=False)
```
Comment (wawltor) [inline examples/biomedical/cblue/train_ner.py]: 这里的magic number是为了和静态图一致吗？

File: examples/biomedical/cblue/utils.py [hunk 7]
```diff
@@ -41,6 +44,52 @@ def create_dataloader(dataset,
...
+        warmup_steps = warmup if is_integer(warmup) else int(
+            math.floor(warmup * total_steps))
+
+        def lr_lambda(current_step):
+            if current_step < warmup_steps:
+                return float(current_step) / float(max(1, warmup_steps))
+            return max(0.0, 1.0 - current_step / total_steps)
+
+        super(LinearDecayWithWarmup, self).__init__(learning_rate, lr_lambda,
+                                                    last_epoch, verbose)
```
Comment (wawltor) [inline examples/biomedical/cblue/utils.py:85]: 如周会讨论，如果对效果影响不大，可以直接使用PaddleNLP的LinearDecayWithWarmup

File: examples/biomedical/cblue/model.py [hunk 8]
```diff
@@ -32,6 +33,7 @@ def __init__(self, electra, num_classes, dropout=None):
         self.classifier_sym = nn.Linear(self.electra.config['hidden_size'],
                                         self.num_classes_sym)
         self.init_weights()
+        self.electra.embeddings.layer_norm._epsilon = self.layer_norm_eps
```
Comment (wawltor) [inline examples/biomedical/cblue/model.py]: 总感觉这样的写法有点丑，我记得是可以直接穿参数到到PretrainedModel里面去

File: examples/biomedical/cblue/README.md (hunk 1 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/README.md:61]: 其他数据集上使用单卡即可。在CHIP-CDN-2C数据集上推荐多卡，否则训练需要约23个小时。

File: examples/biomedical/cblue/train_ner.py (hunk 3 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/train_ner.py:192]: 已修改。

File: examples/biomedical/cblue/train_ner.py (hunk 2 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/train_ner.py]: CBLUEChunkEvaluator可以的。不过SPO任务里也实现了一个Metric，名字设计我们线下再讨论一下吧。

另外，这里还有个问题：
1）PaddleNLP的ChunkEvaluator和现在NEREvaluator的区别是chunk的定义。
2）当前NEREvaluator和EHealth中的实现也有一点diff，num_correct的统计有区别。NEREvaluator在统计真实标签时有去重，EHealth中没有去重，导致NEREvaluator的结果比EHealth中的偏高一点，但与预测结果无关，基本趋势保持一致。这点是不是要严格对齐呢？

File: examples/biomedical/cblue/utils.py (hunk 5 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/utils.py]: 好的，已修改。

File: examples/biomedical/cblue/train_ner.py (hunk 6 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/train_ner.py]: 是的，为了排除attention mask的影响，经排查不影响最后结果，已删除。

File: examples/biomedical/cblue/utils.py (hunk 7 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/utils.py:85]: 目前测试结果表明，使用PaddleNLP的LinearDecayWithWarmup会比当前版本的F1平均低0.02个点(0.66 -> 0.64)。

File: examples/biomedical/cblue/model.py (hunk 8 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/model.py]: 好的，我按照线下讨论的修改一下。

File: examples/biomedical/cblue/README.md (hunk 1 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/README.md:61]: 经线下讨论，暂时保留多卡训练的代码。

File: examples/biomedical/cblue/train_ner.py (hunk 2 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/train_ner.py]: 经线下讨论，两个Metric的名字分别定义为NERChunkEvaluator和SPOChunkEvaluator。

File: examples/biomedical/cblue/utils.py (hunk 7 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/utils.py:85]: 经线下讨论，暂时不改回PaddleNLP中的实现。

File: examples/biomedical/cblue/utils.py (hunk 4 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/utils.py]: Done.

File: examples/biomedical/cblue/train_spo.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+            spo_mask = paddle.matmul(ent_mask, ent_mask, transpose_y=True)
+            spo_mask = paddle.unsqueeze(spo_mask, axis=1)
+
+            with paddle.amp.auto_cast(
+                    args.use_amp,
+                    custom_white_list=['layer_norm', 'softmax', 'gelu'], ):
+                logits = model(input_ids, token_type_ids, position_ids)
+                ent_label, spo_label = create_batch_label(
+                    ent_idx, spo_idx, metric.num_classes, max_batch_len)
+                logits = [F.sigmoid(x) for x in logits]
```
Comment (wawltor) [inline examples/biomedical/cblue/train_spo.py]: 建议使用functional的方式看起来更合适，用partial在这个场景不是特别直观

File: examples/biomedical/cblue/train_spo.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+
+@paddle.no_grad()
+def evaluate(model, criterion, metric, data_loader):
+    """
+    Given a dataset, it evals model and compute the metric.
+    Args:
+        model(obj:`paddle.nn.Layer`): A model to classify texts.
+        dataloader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches.
+        criterion(obj:`paddle.nn.Layer`): It can compute the loss.
```
Comment (wawltor) [inline examples/biomedical/cblue/train_spo.py]: 这里看起来不是paddler.nn.Layer

File: examples/biomedical/cblue/utils.py [hunk 11]
```diff
@@ -132,41 +210,194 @@ def convert_example_ner(example,
...
+    encoded_inputs['position_ids'] = list(range(input_len))
+    encoded_inputs['attention_mask'] = np.ones(input_len)
     if not is_test:
         encoded_inputs['ent_label'] = example['ent_label']
         encoded_inputs['spo_label'] = example['spo_label']
     return encoded_inputs
 
 
-def create_batch_label(ent_labels, spo_labels, num_classes, max_batch_len):
+def create_batch_label(ent_labels,
```
Comment (wawltor) [inline examples/biomedical/cblue/utils.py]: 在模型的计算过程中构建lable，后续会有一个坑，没有办法进行动转静  
可以参考一下BERT这块的写法 https://github.com/PaddlePaddle/PaddleNLP/blob/93cae49c0c572b5c1ac972759140fbe924b0374d/examples/language_model/bert/run_pretrain.py#L184

**Review by ZeyuChen — DISMISSED** [review]
LGTM

File: examples/biomedical/cblue/train_spo.py (hunk 9 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/train_spo.py]: Done.

File: examples/biomedical/cblue/train_spo.py (hunk 10 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/train_spo.py]: 已修改。

**Review by wawltor — APPROVED** [review]
LGTM
